rfcs/0189-contracts.md

```diff
@@ -0,0 +1,630 @@
+_as presented here or a slightly modified if deemed necessary._
+
+[vars]: https://discourse.nixos.org/t/vars-a-framework-for-managing-secrets-and-computed-values/62411
+
+One problem encountered by those modules providing one option per secret
+is the file must be readable by the user of the service.
+This is often solved by relying on [systemd's credentials][] system
+or less securely by using the `root` user in the service startup to read from the file.
+
+[systemd's credentials]: https://systemd.io/CREDENTIALS/
+
+This contract provides an alternative where the `consumer` of the contract - the module requiring a secret - imposes a `user` to the secret `provider`, which here would be [agenix][] or [sops-nix][] for example.
+
+[agenix]: https://github.com/ryantm/agenix
+[sops-nix]: https://github.com/Mic92/sops-nix
+
+Contrary to the previous contracts we covered, the `consumer` here needs to read the `output` of the `provider`
+because it contains the path to the file containing the secret.
+
+When testing a module that expects a file containing a raw secret,
+the ubiquitous method to provide the file is using `pkgs.writeText`.
+This works but has the issue the created file is world readable
+and we thus do not test the file is accessible with the correct user.
+To avoid this pitfall going forwards, we created the `testing.hardcodedSecret`
+`provider` which is an improved version of `pkgs.writeText`
+where the resulting file is created with the requested `owner`, `mode`, etc.
+as described by the contract `consumer`.
+
+This new provider has been tested using the contract' behavior test
+and has been used in `services.stash`'s as an example.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+We are not aware of any because this solution is fully backwards compatible,
+incremental and has a lot advantages. It also arise from a real practical need.
+
+Care should be taken to not abuse this pattern though. It should be reserved
+for contracts where abstracting away a `consumer` and `provider` makes sense.
+We didn't find a general rule for that but a good indication that the pattern gets abused
+is if we only find one `consumer` and `provider` pair in the whole of nixpkgs.
+
+# Alternatives
+[alternatives]: #alternatives
+
+This design arise from trying to minimize code reuse.
+We started by fiddling with nix code and the implementation came up naturally.
+
+We are not aware of any alternatives to do this,
+mostly because our attempts to tweak the code led us often to infinite recursion or other module issues
+so we couldn't stray too far from the way it is written now.
+
+# Prior art
+[prior-art]: #prior-art
+
+We did not find any discussion about any of this by the nix community.
+It is a bit self-centered but the two talks I (ibizaman) gave on this subject in nixpkgs can be considered prior arts.
+Note the syntax presented is a bit outdated now but the message and the passion is still relevant:
```

Comment:
(it wasn't clear to me what you meant by 'passion' here)

Reply:
I mostly meant that passion I had behind this endeavor. :P But I agree it's superfluous.